### DISCUSSION TITLE
😱 Handle Error Gracefully On Function Execution Time Exceeding

### DISCUSSION THREAD

**Discussion by abhagsain** [vercel/next.js#22509] [Help]
I'm using a different API to get data through Next.js API Routes but sometimes the API takes more than 10s and function throws an error if it exceeds that time limit.
The third party API isn't in my control so I can't make it fast. 
So I have come up with this solution. 

```
Promise.race([getData(), sleep(5000)])
    .then((data) => {
        res.statusCode = 200;
        return res.json(data);
    }).catch(() => {
        res.statusCode = 500;
        return res.json({});
    });
```
This way I'm sure that I'll get the response in at most 5s. 

But I'm getting this warning in the console `API resolved without sending a response for /api/data, this may result in stalled requests.` 
I can put some sort of caching to return the old data if the API doesn't respond within 10s. (Would love your suggestions on this)

Is there a good way to handle this?
I'm planning to upgrade to Pro Plan on Vercel which increases the function limit to 60s but I'd still need to handle this case because if the API takes 60s or more then I'll exhaust those resources as well😅

Note: I kinda polling for the new data for every 10s-20s.

**Comment by michrome** [comment]
Is Vercel sending a `SIGTERM` event you can subscribe to? That might be a more conventional way of tidying up before the function is killed. 

https://nodejs.org/api/process.html#process_signal_events

**Reply by abhagsain** [reply]
Hi, Mike. I'm not sure about it. These are the function logs  

![image](https://user-images.githubusercontent.com/36589645/109079475-fe113280-7724-11eb-97c5-1581ead65648.png)

**Reply by michrome** [reply]
I see now that there isn't a `SIGTERM` event, but is [this comment](https://github.com/vercel/next.js/issues/10439#issuecomment-583214126) helpful?

**Comment by heyAyushh** [comment]
if you have bad performance, it can be caused by cold start, which is mostly due to requiring Node.js modules.
 
 Try using https://nextjs.org/docs/basic-features/data-fetching
 
 ```
getStaticProps (Static Generation): Fetch data at build time.
getStaticPaths (Static Generation): Specify dynamic routes to pre-render pages based on data.
getServerSideProps (Server-side Rendering): Fetch data on each request.
 ```
 
 also you can try using https://swr.vercel.app/ to poll and cache data fetching.

**Reply by abhagsain** [reply]
Thanks @heyAyushh 
`getStaticProps, getStaticPath, getServerSideProps` isn't the solution to my problem. 

Here's why 
- I'm using API Routes to bypass CORS in the client side otherwise I could just call the third party API from the client itself. 
- If I use SWR I can get rid of the `setInterval` which I'm currently using but it won't actually solve the problem of **gracefully handling the function limit exceeding error** It's a nice alternative btw. 

Also I'm not sure for how long the function stays up for handling further requests. 
If it creates a new lambda function on every request 😶  then I'm better off having a DO Express/Socket Server with a caching layer on top of it 😄

**Comment by stigkj** [comment]
I am not sure how you have implemented `sleep` shown in your code snippet, but if it is the commonly used

```
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
```

then I would think you get the error because `data` is undefined when the request "times out", i.e. `sleep` will resolve to `undefined` and not throw an exception.
